Sources/BraintreeCoreSwift/BTHTTP.swift

```diff
@@ -0,0 +1,557 @@
+            errorUserInfo["method"] = method
+            errorUserInfo["path"] = path
+            errorUserInfo["parameters"] = parameters
+
+            let error = constructError(code: .missingBaseURL, userInfo: errorUserInfo)
+
+            completion(nil, error)
+            return
+        }
+        
+        let fullPathURL: URL?
+        let isNotDataURL: Bool = baseURL.scheme != "data"
+
+        if isNotDataURL, let path = path {
+            fullPathURL = hasHTTPPrefix ? URL(string: path) : baseURL.appendingPathComponent(path)
+        } else {
+            fullPathURL = baseURL
+        }
+
+        let mutableParameters: NSMutableDictionary = NSMutableDictionary(dictionary: parameters ?? [:])
+
+        if authorizationFingerprint != "" {
+            mutableParameters["authorization_fingerprint"] = authorizationFingerprint
+        }
+
+        guard let fullPathURL = fullPathURL else {
+            /// baseURL can be non-nil (e.g. an empty string) and still return nil for -URLByAppendingPathComponent:
+            /// causing a crash when NSURLComponents.componentsWithString is called with nil.
+            errorUserInfo["method"] = method
+            errorUserInfo["path"] = path
+            errorUserInfo["parameters"] = parameters
+            errorUserInfo[NSLocalizedFailureReasonErrorKey] = "fullPathURL was nil"
+
+            let error = constructError(code: .missingBaseURL, userInfo: errorUserInfo)
+
+            completion(nil, error)
+            return
+        }
+
+        buildHTTPRequest(
+            method: method,
+            url: fullPathURL,
+            parameters: mutableParameters,
+            isNotDataURL: isNotDataURL
+        ) { request, error in
+            completion(request, error)
+        }
+    }
+
+    func buildHTTPRequest(
+        method: String?,
+        url: URL,
+        parameters: NSMutableDictionary? = [:],
+        isNotDataURL: Bool,
+        completion: @escaping (URLRequest?, Error?) -> Void
+    ) {
+        guard var components: URLComponents = URLComponents(string: url.absoluteString) else {
+            let error = constructError(
+                code: .urlStringInvalid,
+                userInfo: [NSLocalizedDescriptionKey: "The URL absolute string is malformed or invalid."]
+            )
+
+            completion(nil, error)
+            return
+        }
+
+        var headers: [String: String] = defaultHeaders()
+        var request: URLRequest
+
+        if method == "GET" || method == "DELETE" {
+            if isNotDataURL {
+                components.percentEncodedQuery = BTURLUtils.queryString(from: parameters ?? [:])
+            }
+            guard let urlFromComponents = components.url else {
+                let error = constructError(
+                    code: .urlStringInvalid,
+                    userInfo: [NSLocalizedDescriptionKey: "The URL absolute string is malformed or invalid."]
```

Comment:
New errors that we add should be included in the CHANGELOG

Reply:
If it's merchant facing*^. Which, we can double check if this one would be, but I am _assuming_ we bubble this up to the merchant

Reply:
Addressed in this comment - https://github.com/braintree/braintree_ios/pull/855#discussion_r955000739